identify/identify.py

```diff
@@ -25,12 +25,16 @@
 TEXT = 'text'
 BINARY = 'binary'
 
-ALL_TAGS = {DIRECTORY, SYMLINK, FILE, EXECUTABLE, NON_EXECUTABLE, TEXT, BINARY}
-ALL_TAGS.update(*extensions.EXTENSIONS.values())
-ALL_TAGS.update(*extensions.EXTENSIONS_NEED_BINARY_CHECK.values())
-ALL_TAGS.update(*extensions.NAMES.values())
-ALL_TAGS.update(*interpreters.INTERPRETERS.values())
-ALL_TAGS = frozenset(ALL_TAGS)
+TYPE_TAGS = frozenset({DIRECTORY, FILE, SYMLINK})
+MODE_TAGS = frozenset({EXECUTABLE, NON_EXECUTABLE})
+ENCODING_TAGS = frozenset({BINARY, TEXT})
+FORMAT_TAGS = frozenset().union(
+    *extensions.EXTENSIONS.values(),
+    *extensions.EXTENSIONS_NEED_BINARY_CHECK.values(),
+    *extensions.NAMES.values(),
+    *interpreters.INTERPRETERS.values(),
+) - ENCODING_TAGS
+ALL_TAGS = frozenset({*TYPE_TAGS, *MODE_TAGS, *ENCODING_TAGS, *FORMAT_TAGS})
```

Comment:
this does a lot of copies that weren't present in the left hand side -- can you validate that it doesn't make import time significantly slower for this module?

Reply:
actually, this can be done much simpler by just adding your new constants after the old code instead of modifying how the current set of `ALL_TAGS` is built, then unbuilt, then built

Reply:
Running a simple script that tests the time to import `identify.identify` with `python <testscript.py> in a conda environment with Python 3.7.9 and the respective dev versions of identify isntalled with `pip install -e .` over 12 runs, discarding the first two:

Import time, this PR: 30.6 ms
Import time, `master`: 30.9 ms

A small advantage to this PR, but well within the variability of the test runs.

Measuring the `time.time()` elapsed running the actual code block modified in `identify.identify`, averaged over 1000 runs in a loop:

Prior to PR: 39.4 us
With PR: 61.4 us

The same, but placed in a function and measured with IPython `%timeit`:

Prior to PR: 41.8 us +/- 2.08 us
With PR: 63.3 us +/- 1.18 us

Test script:

```python
import time
t1 = time.time()
import identify.identify
t2 = time.time()
delta_t = t2 - t1
print(delta_t)
```

Reply:
To give some fun context to the above results, assuming the ≈20 us extra time cost per import suggested above, it will take around 100 million Identify imports to equal the ≈30 minutes it took me to set up, execute, tabulate and write up the above tests.

Reply:
you're not really profiling properly, imports only execute the first time they're run

but also it isn't what we agreed upon in the original issue -- adding constants, not rewriting the current code (as much as I know you love rewriting things!)

Reply:
Sorry, for some reason Github wasn't showing your other comments on this PR until I force-reloaded.

> actually, this can be done much simpler by just adding your new constants after the old code instead of modifying how the current set of `ALL_TAGS` is built, then unbuilt, then built

> but also it isn't what we agreed upon in the original issue -- adding constants, not rewriting the current code

Actually, I originally implemented this using the exact code snippit I'd proposed in #157 (aside from bug fixes) that I thought was what you had said you were okay with. Regardless, your suggestion makes much more sense, and I went back and restored the previous `ALL_TAGS` building process, just adding the hardcoded tag group constants before and the format tag group after. As you expected, this eliminates most of the small performance hit from this PR, with the code block now taking only 48.3 us to execute, vs. 61.4 us before, only 6.5 us slower than the previous 41.8 us, and nearly 4 OoM less than the import time (30 ms).

Also, I've rebased to separate each atomic change into a separate commit for easier review, and so I can drop any (e.g. change to use the constants rather than hardcoding in the tests) if not desired.

> you're not really profiling properly, imports only execute the first time they're run

As described above, the benchmarks reported involve either starting a fresh interpreter every iteration, or measuring the execution time of just running the specific code block in question (since run to run variability and the precision of my machine's OS timers for the former greatly outweighs the latter performance impact). Neither is subject to the import system's caching of previously loaded modules; I'm not importing in a for loop or in `%timeit` inside the same interpreter or anything silly like that, which would result in a time >4 OoM smaller than reported above (on the order of 1 us, not 30 ms). If there's some other way you'd like me to profile it that I haven't already, please let me know.

> (as much as I know you love rewriting things!)

Yeah, most scientists don't spend enough time refactoring their code, but I think sometimes I take it a _little_ too far in the opposite direction...